src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs

```diff
@@ -1242,92 +1277,127 @@ internal static bool DeepEquals(JsonElement left, JsonElement right)
                     return true;
 
                 case JsonValueKind.Number:
-                    return JsonHelpers.AreEqualJsonNumbers(left.GetRawValue().Span, right.GetRawValue().Span);
+                    return JsonHelpers.AreEqualJsonNumbers(element1.GetRawValue().Span, element2.GetRawValue().Span);
 
                 case JsonValueKind.String:
-                    if (right.ValueIsEscaped)
+                    if (element2.ValueIsEscaped)
                     {
-                        if (left.ValueIsEscaped)
+                        if (element1.ValueIsEscaped)
                         {
                             // Both values are escaped, force an allocation to unescape the RHS.
-                            return left.ValueEquals(right.GetString());
+                            return element1.ValueEquals(element2.GetString());
```

Comment:
There's no way to do this without allocating? e.g. unescaping into a rented buffer?

Reply:
I think it would require a new unescaping routine but I'll see if I can knock together something something simpler.

Reply:
I've added a new routine that checks if two escaped strings are equal.